### PR TITLE
WIP: Testing Telegram Update Handler

### PR DESCRIPTION
**Pull request by 10eMyrT**
This PR tests the update handler of Telegram. Just the functionalities of the update handler are tested, and what is done by the functions called in this handler is not addressed because each of those handlers have already been tested separately.

### CODE REVIEW

File: internal/handlers/telegram/handler_test.go [hunk 1]
```diff
@@ -10,6 +10,160 @@ import (
...
+			return "documentHandler triggered"
+		case u.Message.Photo != nil:
+			return "photoHandler triggered"
+		default:
+			return "Triggered, but message type is currently unsupported"
+		}
+	}
+	return ""
+}
+
```
Comment (10eMyrT) [inline internal/handlers/telegram/handler_test.go:36]: I had to create this function here based on the handler itself, and remove the calls to the other methods to make it work.

File: internal/handlers/telegram/handler_test.go [hunk 2]
```diff
@@ -10,6 +10,160 @@ import (
...
+						FirstName: "test",
+						UserName:  "testUser",
+					},
+				},
+			},
+		}
+	updates <- update
+
+	correct := "joinHandler triggered"
+	assert.Equal(t, correct, MockUpdateHandler(updates))
```
Comment (10eMyrT) [inline internal/handlers/telegram/handler_test.go:54]: This strategy works, and running `go test ./...` does complete successfully.

File: internal/handlers/telegram/handler_test.go [hunk 3]
```diff
@@ -10,6 +10,160 @@ import (
...
+	// 		Suffix:          ">",
+	// 		ShowJoinMessage: true,
+	// 	},
+	// 	logger: internal.Debug {
+	// 		DebugLevel: true,
+	// 	},
+	// 	sendToIrc: func(s string) {
+	// 	},
+	// }
+	// updateHandler(clientObj, updates)
```
Comment (10eMyrT) [inline internal/handlers/telegram/handler_test.go:68]: Calling the update handler takes long time, and does not even complete when I run `go test ./...`. Your input on the reason is welcome, but I believe it stays in the loop waiting for another update.

Comment (Tjzabel) [reply internal/handlers/telegram/handler_test.go:68]: Most likely due to waiting for update events.

File: internal/handlers/telegram/handler_test.go (hunk 1 above)
Comment (robbyoconnor) [reply internal/handlers/telegram/handler_test.go:36]: This looks incredibly fragile and prone to code rot if people forget this happened...

Comment (jwflory) [reply internal/handlers/telegram/handler_test.go:36]: @robbyoconnor Could you please clarify what you mean by if people forget this happens? This is in the context of unit tests, right?

Comment (robbyoconnor) [reply internal/handlers/telegram/handler_test.go:36]: I withdraw my comment.

File: internal/handlers/telegram/handler_test.go (hunk 3 above)
Comment (Tjzabel) [reply internal/handlers/telegram/handler_test.go:68]: @Zedjones do you know why the client would be stalling on the channel here?

**Comment by jwflory** [comment]
@Tjzabel @Zedjones @10eMyrT Is this PR in scope for the release today? Trying to decide if I bump to the `v2.x.x` milestone or keep it on target for today.

**Comment by Zedjones** [comment]
I think a better way to go about this might be to use `var` at the top level of the module to mock out each handler individually. This current method does not allow for protections against regressions when adding functionality to the `updateHandler` and it also doesn't account for the logging being done in the actual `updateHandler`.

[Here's](https://github.com/Zedjones/RedirectProtect/blob/3d20fc6a012a04e7f4ffab810163a562e9b6e878/internal/time_checking.go) an example of how to use `var` at the top level. And [here's](https://github.com/Zedjones/RedirectProtect/blob/3d20fc6a012a04e7f4ffab810163a562e9b6e878/internal/time_checking_start_time_check_test.go) an example of how to use that in a test.

**Comment by jwflory** [comment]
_Discussed in [2020-05-23 meeting](https://bluejeans.com/s/kF2he)._

---

## Release-blocking: _No_

We discussed in the meeting, and @Zedjones summarized the meeting feedback in his comments above. We decided we won't try to rush or expedite these tests for the release if they are not fully polished yet.

Will follow up after v2.0.0 release.

**Comment by 10eMyrT** [comment]
Thanks for the reviews. I will get back to it this week and push an update as soon as finished.

**Comment by Tjzabel** [comment]
Closing due to inactivity.

**Pull request closed without merging**